src/client/api/object.c

```diff
@@ -440,3 +440,368 @@ daos_oit_list(daos_handle_t oh, daos_obj_id_t *oids, uint32_t *oids_nr,
 	D_FREE(oa);
 	return rc;
 }
+
+struct oit_mark_arg {
+	void		*oma_buf;
+	d_sg_list_t	*oma_sgl;
+	daos_iod_t	*oma_iod;
+	daos_obj_id_t	 oma_oid;
+	daos_key_t	 oma_dkey;
+	uint32_t	 oma_bid;
+};
+
+static void
+oit_mark_arg_free(struct oit_mark_arg *arg)
+{
+	if (arg == NULL)
+		return;
+	D_FREE(arg);
+}
+
+static struct oit_mark_arg *
+oit_mark_arg_alloc(daos_obj_id_t oid, d_iov_t *marker)
+{
+	struct oit_mark_arg	*arg;
+	void			*buf, *ptr, *val;
+	d_iov_t			*iov;
+	size_t			 val_len, buf_len;
+
+	val_len = DAOS_OIT_MARKER_MAX_LEN + DAOS_OIT_DEFAULT_VAL_LEN;
+	buf_len = roundup(sizeof(struct oit_mark_arg), 8) + sizeof(d_sg_list_t) + sizeof(d_iov_t) +
+		  roundup(sizeof(daos_iod_t), 8) + roundup(val_len, 8);
+
+	D_ALLOC(buf, buf_len);
+	if (buf == NULL)
+		return NULL;
+
+	ptr = buf;
+	arg = (struct oit_mark_arg *)ptr;
+	arg->oma_buf = buf;
+	arg->oma_oid = oid;
+	arg->oma_bid = 0;
+	DAOS_OIT_DKEY_SET(&arg->oma_dkey, &arg->oma_bid);
+	ptr += roundup(sizeof(struct oit_mark_arg), 8);
+	arg->oma_sgl = (d_sg_list_t *)ptr;
+	ptr += sizeof(d_sg_list_t);
+	iov = (d_iov_t *)ptr;
+	ptr += sizeof(d_iov_t);
+	arg->oma_iod = (daos_iod_t *)ptr;
+	ptr += roundup(sizeof(daos_iod_t), 8);
+	val = ptr;
+
+	arg->oma_sgl->sg_nr = arg->oma_sgl->sg_nr_out = 1;
+	arg->oma_sgl->sg_iovs = iov;
+	if (marker != NULL) {
+		D_ASSERT(marker->iov_len <= DAOS_OIT_MARKER_MAX_LEN);
+		memcpy(val + DAOS_OIT_DEFAULT_VAL_LEN, marker->iov_buf, marker->iov_len);
+		d_iov_set(iov, val, DAOS_OIT_DEFAULT_VAL_LEN + marker->iov_len);
+	} else {
+		d_iov_set(iov, val, DAOS_OIT_DEFAULT_VAL_LEN);
+	}
+
+	DAOS_OIT_AKEY_SET(&arg->oma_iod->iod_name, &arg->oma_oid);
+	arg->oma_iod->iod_type	= DAOS_IOD_SINGLE;
+	arg->oma_iod->iod_size	= iov->iov_len;
+	arg->oma_iod->iod_nr	= 1;
+
+	return arg;
+}
+
+static int
+oit_mark_cb(tse_task_t *task, void *data)
+{
+	struct oit_mark_arg	*arg = *((struct oit_mark_arg **)data);
+
+	oit_mark_arg_free(arg);
+	return 0;
+}
+
+int
+daos_oit_mark(daos_handle_t oh, daos_obj_id_t oid, d_iov_t *marker, daos_event_t *ev)
```

Comment:
please add some comments or README.md on how the marker is stored (the underlying format)

Reply:
sure.